worker-message/src/worker.rs

```diff
@@ -0,0 +1,325 @@
+use crate::message::{ExecuteOutput, Request, Response};
+use crate::{
+    CoordinatorMessage, ExecuteCommandRequest, ExecuteCommandResponse, JobReport, ReadFileRequest,
+    ReadFileResponse, ResponseResult, StreamCommandRequest, StreamCommandResponse, WorkerMessage,
+    WriteFileRequest, WriteFileResponse,
+};
+use std::path::{Path, PathBuf};
+use std::process::{Output, Stdio};
+
+use anyhow::{Error, Result};
+use std::time::SystemTime;
+use tempdir::TempDir;
+use tokio::fs;
+use tokio::io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader};
+use tokio::process::{Child, Command};
+use tokio::sync::mpsc;
+use tokio::task::{JoinHandle, JoinSet};
+use tracing::info;
+
+pub async fn listen() {
+    let mut tasks = JoinSet::new();
+    let (tx, mut coordinator_rx) = mpsc::channel(8);
+    tasks.spawn(async move {
+        let mut buffer = Vec::new();
+        let mut input = tokio::io::stdin();
+        loop {
+            let payload_len = input
+                .read_u64()
+                .await
+                .expect("Failed to read standard input");
+            buffer.resize(payload_len as usize, 0);
+            input
+                .read_exact(&mut buffer)
+                .await
+                .expect("Failed to read standard input");
+            let coordinator_msg: CoordinatorMessage =
+                bincode::deserialize(&buffer).expect("Failed to deserialize CoordinatorMessage");
+            tx.send(coordinator_msg).await.ok(/*Only fail due to process exiting*/);
+        }
+    });
+    let (coordinator_tx, mut rx) = mpsc::channel(8);
+    tasks.spawn(async move {
+        let mut output = tokio::io::stdout();
+        loop {
+            let worker_msg = rx.recv().await.expect("Failed to write to standard output");
+            info!("{:#?}", worker_msg);
+            let encoded =
+                bincode::serialize(&worker_msg).expect("Failed to serialize WorkerMessage");
+            output
+                .write_u64(encoded.len() as u64)
+                .await
+                .expect("Failed to write to standard output");
+            output
+                .write_all(&encoded)
+                .await
+                .expect("Failed to write to standard output");
+            output
+                .flush()
+                .await
+                .expect("Failed to flush standard output");
```

Comment:
This is much more compact. I'm still not quite sure about using SNAFU since many errors in `worker` are unrecoverable.

Reply:
"unrecoverable" Is a tricky thing. This specific error may indeed be unrecoverable, but I'd expect most of the errors in the worker can occur but then other work may continue. Making up something, perhaps starting `cargo fmt` would fail, but we should still be able to subsequently to attempt `cargo build`.

There's also the issue of error detail. For example, the IIFE version makes it so we cannot distinguish between the three failure cases. The extreme form of grouping errors like this is to have a single "some error occurred" message, which as you can imagine isn't very useful. 

Even if we cannot recover from the errors, presumably we'd want to perform some kind of external system maintenance to address the problem. In general, the more detail we have, the more chance we have to figure out the problem.

All of that (and probably more) is why I think there's a bit of an art to figuring out what detail is useful to capture 😉 .